retryrpc/api.go

```diff
@@ -0,0 +1,134 @@
+// Package retryrpc provides a client and server RPC model which survives
+// lost connections on either the client or the server.
+package retryrpc
+
+import (
+	"container/list"
+	"sync"
+	"time"
+)
+
+// Request is the structure sent
+type Request struct {
+	Len       uint64 // Length of this struct including len field
+	RequestID uint64 // Unique ID of this request
+	MountID   uint64 // Mount ID
+	JReq      []byte // JSON containing request
+}
+
+// Response is the structure returned
+type Response struct {
+	Len uint64 // Length of this struct including len field
+	Res []byte // JSON response - only valid if err != nil
+	Err error  // Status of request
+}
+
+type completedRequest struct {
+	completedTime time.Time
+	request       *Request
+}
+
+// Server tracks the state of the server
+type Server struct {
+	sync.Mutex
+	completedTTL      time.Duration // How long a completed request stays on queue
+	serviceMap        map[string]*func(request interface{}, response interface{})
+	connections       *list.List                   // List of outstanding connections
+	pendingRequests   map[string]*Request          // Request which have not been completed yet.
```

Comment:
For maps, I like to be told at least what the Key is... I can (usually) see what the Value is :-).

As for the Key being a `string`, I'm actually surprised at this. Perhaps you do something with RequestID and MountID to be the string that is the Key here? There are more direct ways to do that.

For instance, you can look at type `readCacheKeyStruct` in inode/config.go:
```
type readCacheKeyStruct struct {
	volumeName       string
	logSegmentNumber uint64
	cacheLineTag     uint64 // LogSegment offset / readCacheLineSize
}
```
It is used as the Key in the readCache field of the volumeGroupStruct listed just below.
In this case, since it's a named Key type, I didn't bother commenting what it is :-).

Anyway, this is just a cute thing you can do with Go maps I thought more appropriate here.

Reply:
Good catch.   The key for pendingRequests will be a combination of mountID and client request ID.   I added this to the code:

// RequestKey uniquely identifies a client request from a client
type RequestKey struct {
	uniqueID string
	clientRequestID uint64
}